File: src/main/java/com/github/windchopper/tools/log/browser/fs/RemoteFile.java
package com.github.windchopper.tools.log.browser.fs;

import javafx.beans.property.BooleanProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.scene.control.ListView;
import javafx.scene.control.TreeView;
import org.apache.commons.lang3.StringUtils;

import java.util.Map;

public abstract class RemoteFile implements Comparable<RemoteFile> {

    private boolean selected;

    public boolean isSelected() {
        return selected;
    }

    public void setSelected(boolean selected) {
        this.selected = selected;
    }

    public abstract String path();
    public abstract boolean directory();

    @Override public int compareTo(RemoteFile anotherFile) {
        int result = directory() == anotherFile.directory() ? 0 : directory() ? -1 : +1;

        if (result != 0) {
            return result;
        }

        String name = StringUtils.substringAfterLast(path(), "/");
        String anotherName = StringUtils.substringAfterLast(anotherFile.path(), "/");

        result = StringUtils.equals(name, "..") == StringUtils.equals(anotherName, "..") ? 0 : StringUtils.equals(name, "..") ? -1 : +1;

        if (result != 0) {
            return result;
        }

        return name.compareTo(anotherName);
    }

    public String displayName(boolean appendDirectoryNameWithSlash) {
        String displayName = StringUtils.defaultString(StringUtils.trimToNull(StringUtils.substringAfterLast(path(), "/")), "/");

        if (directory() && appendDirectoryNameWithSlash) {
            displayName += "/";
        }

        return displayName;
    }

    public BooleanProperty createSelectedProperty(Map<String, BooleanProperty> selectedStateBuffer, TreeView<?> treeView, ListView<?> listView) {
        BooleanProperty selectedProperty = selectedStateBuffer.computeIfAbsent(path(), missingPath -> new SimpleBooleanProperty(this, "selected"));

        selectedProperty.addListener((observable, oldValue, newValue) -> {
            treeView.refresh();
            listView.refresh();
        });

        return selectedProperty;
    }

}

File: src/main/java/com/github/windchopper/tools/log/browser/fx/RemoteFile.kt
package com.github.windchopper.tools.log.browser.fx

import javafx.beans.property.BooleanProperty
import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.value.ObservableValue
import javafx.scene.control.ListView
import javafx.scene.control.TreeView
import org.apache.commons.lang3.StringUtils
import java.nio.file.Files
import java.nio.file.Path

@Suppress("UNUSED_PARAMETER", "UNUSED_ANONYMOUS_PARAMETER") class RemoteFile(val path: Path): Comparable<RemoteFile> {

    val directory: Boolean
        get() = Files.isDirectory(path)

    override fun compareTo(other: RemoteFile): Int {
        return path.compareTo(other.path)
    }

    fun displayName(appendDirectoryNameWithSlash: Boolean): String {
        var displayName = StringUtils.defaultString(StringUtils.trimToNull(StringUtils.substringAfterLast(path.toString(), "/")), "/")
        if (directory && appendDirectoryNameWithSlash) {
            displayName += "/"
        }
        return displayName
    }

    fun createSelectedProperty(selectedStateBuffer: MutableMap<String, BooleanProperty>, treeView: TreeView<*>, listView: ListView<*>): BooleanProperty {
        val selectedProperty = selectedStateBuffer.computeIfAbsent(path.toString()) { SimpleBooleanProperty(this, "selected") }
        selectedProperty.addListener { observable: ObservableValue<out Boolean?>?, oldValue: Boolean?, newValue: Boolean? ->
            treeView.refresh()
            listView.refresh()
        }
        return selectedProperty
    }

}